code/datums/components/uplink.dm

```diff
@@ -170,94 +148,106 @@
-			var/datum/uplink_item/I = uplink_items[category][item]
-			if(I.limited_stock == 0)
-				continue
-			if(length(I.restricted_roles))
-				if(!debug && !(user.mind.assigned_role.title in I.restricted_roles))
-					continue
-			if(I.restricted_species)
-				if(ishuman(user))
-					var/is_inaccessible = TRUE
-					var/mob/living/carbon/human/H = user
-					for(var/F in I.restricted_species)
-						if(F == H.dna.species.id || debug)
-							is_inaccessible = FALSE
-							break
-					if(is_inaccessible)
-						continue
-			cat["items"] += list(list(
-				"name" = I.name,
-				"cost" = I.cost,
-				"desc" = I.desc,
-			))
-		data["categories"] += list(cat)
+	data["uplink_flag"] = uplink_handler.uplink_flag
+	data["has_progression"] = uplink_handler.has_progression
+	data["has_objectives"] = uplink_handler.has_objectives
+	data["lockable"] = lockable
+	data["assigned_role"] = uplink_handler.assigned_role
+	data["debug"] = uplink_handler.debug_mode
 	return data
 
-/datum/component/uplink/ui_act(action, params)
+/datum/component/uplink/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/json/uplink)
+	)
+
+/datum/component/uplink/ui_act(action, params, datum/tgui/ui, datum/ui_state/state)
 	. = ..()
 	if(.)
 		return
 	if(!active)
 		return
 	switch(action)
 		if("buy")
-			var/item_name = params["name"]
-			var/list/buyable_items = list()
-			for(var/category in uplink_items)
-				buyable_items += uplink_items[category]
-			if(item_name in buyable_items)
-				var/datum/uplink_item/I = buyable_items[item_name]
-				MakePurchase(usr, I)
-				return TRUE
+			var/datum/uplink_item/item_path = text2path(params["path"])
+			if(!ispath(item_path, /datum/uplink_item))
+				return
+
+			var/datum/uplink_item/item = GLOB.uplink_items_by_type[item_path]
+			uplink_handler.purchase_item(ui.user, item)
 		if("lock")
 			active = FALSE
 			locked = TRUE
-			telecrystals += hidden_crystals
-			hidden_crystals = 0
 			SStgui.close_uis(src)
-		if("select")
-			selected_cat = params["category"]
-			return TRUE
-		if("compact_toggle")
-			compact_mode = !compact_mode
+
+	if(!uplink_handler.has_objectives)
+		return TRUE
+
+	switch(action)
+		if("regenerate_objectives")
+			uplink_handler.generate_objectives()
 			return TRUE
 
-/datum/component/uplink/proc/MakePurchase(mob/user, datum/uplink_item/U)
-	if(!istype(U))
-		return
-	if (!user || user.incapacitated())
-		return
-	if(U.restricted_roles.len && !(user.mind.assigned_role.title in U.restricted_roles))
-		return
+	var/list/objectives
+	switch(action)
+		if("start_objective")
+			objectives = uplink_handler.potential_objectives
+		if("objective_act", "finish_objective", "objective_abort")
+			objectives = uplink_handler.active_objectives
 
-	if(telecrystals < U.cost || U.limited_stock == 0)
+	if(!objectives)
 		return
-	telecrystals -= U.cost
 
-	U.purchase(user, src)
+	var/objective_index = text2num(params["index"])
+	if(objective_index < 1 || objective_index > length(objectives))
```

Comment:
Make sure to `floor` or similar, this will still allow decimals through, leading `objective` to be null.